Document the possibility of high-level external calls to precompiled contracts

mattaereal opened the pull request:
Hi! I'm trying to add a bit more clarity about something that has been somewhat recently documented but wasn't as easy as I expected to understand from the docs.

A few weeks ago, while checking out the code of Blast (L2), we learned that you can do external calls over precompiled contracts without failing and were wondering how that could be possible. After debugging the compiler and, as the suggestion states, from v0.8.10, the extcodesize check is removed from external calls if there's return data to decode, since it would fail anyway at the time of decoding.

We thought this behavior should be noted on different occasions, particularly while working with precompiled contracts (we thought it wouldn't work). So, without modifying much and reusing what was already written (in the docs and in the commit message), I added a few references and included a note that was outside a callout inside of another.

mattaereal committed:
Adding clarity on extcalls with precompiled contracts.

cameel commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -600,3 +600,8 @@ precompiled contracts. It might also be possible that new
 ``1`` and ``0xffff`` (inclusive).
+
+.. warning::
+    Since the v[0.8.10](https://github.com/ethereum/solidity/commit/a1aa9d2d90f2f7e7390408e9005d62c7159d4bd4) release the compiler does not check `extcodesize` on external calls if return data is expected, since an empty code will be unable to return data, and the ABI decoder will revert. This is an exception for external calls over precompiled contracts which are able to return data without code ...
```
Some remarks on our docs conventions and syntax:
- Note that this file is RestructuredText, not Markdown. Some syntax differs, like links or double backticks.
- We don't normally link to the release in those notes. Not sure it adds anything, especially if it's just pointing at the commit. If there was a good reason to do it, I'd link to the release post on the blog.
- Please break long lines. Best after each sentence.

```suggestion
Since the 0.8.10 release the compiler does not check ``extcodesize`` on external calls if return data is expected,
since an empty code will be unable to return data, and the ABI decoder will revert.
This is an exception for external calls over precompiled contracts which are able to return data without code associated with their addresses.
```

cameel reviewed (changes requested):
Thanks for your interest in improving the docs!

But well, this isn't an exception made on purpose for precompiles. Looking back at the PR that introduced it, this is technically an unwanted change in behavior, but a pretty obscure one and not harmful so it was not considered worth addressing or documenting in detail (see e.g. https://github.com/ethereum/solidity/pull/12205#discussion_r740989862). Before 0.8.10 a high-level external call to a precompiled contract would revert due to that check, but it did not really matter, since they're pretty low-level themselves and we expect people to use low-level calls for that.

And TBH these details seem way too obscure to put them in "Introduction to Smart Contracts". That spot in `control-structures.rst` would be more appropriate. You could just add a sentence there saying that the consequence of this is that it makes high-level external calls to precompiles possible.

TBH, I'm not sure we should document it though. If we document it, we commit to officially supporting it and to me it seems more like a grey area of undefined behavior that's not guaranteed. Though if you connect the dots based on the current docs, you can already conclude that it means that high-level calls to precompiles should work so maybe it does not matter - we already stepped into that one.

tinchoabbate commented:
Hi there! Adding my perspective to the discussion:

>  since they're pretty low-level themselves and we expect people to use low-level calls for that.

AFAIK, that's not happening in reality. Some people might be using low-level calls as you expect, but there're others using this behavior of the compiler. For example, to perform high-level calls to custom precompiles in L2s.

> TBH, I'm not sure we should document it though. If we document it, we commit to officially supporting it and to me it seems more like a grey area of undefined behavior that's not guaranteed.

By documenting it we'd describe the actual behavior of the compiler. I mean, it _already_ works like this. So what's the point in hiding it? If the team doesn't want to officially support it, then the compiler's code should be changed, right? If it's a bug it should be fixed, if it's a feature it should be documented clearly. I find it hard to see a gray area.

> Though if you connect the dots based on the current docs, you can already conclude that it means that high-level calls to precompiles should work so maybe it does not matter - we already stepped into that one.

Documenting it does matter. It removes uncertainty from developers using the compiler and from us security folks who dive into these kind of details of Solidity. I'd rather have it documented explicitly than somehow becoming an obscure behavior of the compiler that developers can run into unexpectedly. It's best to have a clear documentation that people can rely on instead of on unsafe assumptions.

mattaereal commented:
Weird. The tests that are failing aren't related to the documentation.

cameel commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -600,3 +600,10 @@ precompiled contracts. It might also be possible that new
+.. warning::
+    Since the 0.8.10 release the compiler does not check ``extcodesize`` on external calls if return data is expected,
+    since an empty code will be unable to return data, and the ABI decoder will revert.
+    This is an exception for external calls over precompiled contracts which are able to return data without code associated with their addresses.
```
I still think that "Introduction to Smart Contracts" is not the right place for this kind of detail. It would be much better off added to the note about precompiles that we alredy have under "External Function Calls". It would be enough to have a link to that section here.

Also, it should make a clear distinction between *low-level* and *high-level* external calls. I actually found it pretty confusing without that distinction when I was reading it for the first time.

mattaereal replied:
I'm on it. Though I'm not on my setup so I'll be doing it via web browser.

cameel reviewed (commented):
I discussed this with @ekpyron and we're fine with this being documented after all. Like I said, the feature is mostly unintended, but we don't see any good reason to change this behavior in the future so we can just as well explain how it works, as long as it's done in a way that won't be more confusing than helpful (it's a pretty obscure corner case after all).

cameel commented:
> I mean, it already works like this. So what's the point in hiding it?

The point is that some details fall into the category of [unspecified behavior](https://en.wikipedia.org/wiki/Unspecified_behavior). A spec does not always prescribe compiler's behavior in minute detail and things that are not in the spec are subject to change between versions, implementations, etc. For us the docs are the closest thing to a spec that we have. So we sometimes intentionally don't describe every detail, because we don't want users to rely on it.

In this case we decided we're actually fine making this a part of the spec, but the point still stands in general. I mean, would be best to have both - a spec and a detailed description of how our implementation works, but we're a bit resource constrained at the moment so the best we can manage are the docs in the current form.

cameel commented:
> Weird. The tests that are failing aren't related to the documentation.

You can ignore them. This is a documentation change and those tests can't possibly be affected by it.

Our external tests are notoriously flaky unfortunately. They're just tests suites of several real projects written in Solidity that we use to verify backwards-compatibility of changes to the compiler and for benchmarking. Unfortunately the JS ecosystem is a dependency hell and those test suites are not always deterministic so they break randomly all the time. They will probably succeed if you rerun them, but they're not required to merge the PR anyway.

mattaereal commented:
I think I fulfilled what you requested @cameel.
1. Removed the warning on precompiled contracts from introductory since it's pretty obscure for an introduction and may change in the future (idk).
2. Added a note on control-structure file in the section of external calls.
3. Specifically mentioned high-level calls

nikola-matic commented on docs/control-structures.rst:
```diff
@@ -111,22 +111,38 @@ otherwise, the ``value`` option would not be available.
+    contracts possible, since they can return data despite having no code
+    associated with their addresses.
+
+    Read about :ref:`precompiled contracts <precompiledContracts>` and 
```
```suggestion
    Read about :ref:`precompiled contracts <precompiledContracts>` and
```

Check style failing due to this.

nikola-matic commented on docs/control-structures.rst:
```diff
@@ -111,22 +111,38 @@ otherwise, the ``value`` option would not be available.
+    :ref:`low-level calls <address_related>`
+    for more information.
+
+
+
+
```
```suggestion
```

Please remove the superfluous newlines.

nikola-matic commented on docs/control-structures.rst:
```diff
@@ -111,22 +111,38 @@ otherwise, the ``value`` option would not be available.
     above logic even though they execute code and can return data.
 
+.. note::
+    Since the 0.8.10 release the compiler does not check ``extcodesize`` on
```
```suggestion
    Since the version 0.8.10, the compiler does not check ``extcodesize`` on
```

nikola-matic reviewed (commented):
Heya @mattaereal, can you please rebase, address the comments, and squash the commits down to one, so that we can hopefully merge this.

mattaereal commented:
I'll rebase and squash the commits down to one when I get to my computer (currently traveling)

cameel commented on docs/control-structures.rst:
```diff
@@ -111,22 +111,35 @@ otherwise, the ``value`` option would not be available.
+    high-level external calls if return data is expected, because an empty code
+    will be unable to return data, and the ABI decoder will revert.
+    As a consequence, this allows high-level external calls to precompiled
+    contracts possible, since they can return data despite having no code
```
```suggestion
    contracts, since they can return data despite having no code
```

cameel commented on docs/control-structures.rst:
```diff
@@ -900,3 +913,4 @@ in scope in the block that follows.
     The caller always retains at least 1/64th of the gas in a call and thus
     even if the called contract goes out of gas, the caller still
     has some gas left.
+    
```
[`chk_coding_style` is failing here due to trailing whitespace](https://app.circleci.com/pipelines/github/ethereum/solidity/37019/workflows/87b2195e-5c4b-4670-9e37-7f958e47c9ab/jobs/1693393). Is the empty line even necessary here?

mattaereal replied:
No, feel free to remove it. Probably the IDE I used automatically added it somehow.

cameel reviewed (dismissed):
The content itself looks fine now so I'm approving, but there are two style issues that need to be fixed before we can merge. One of them is making a require CI job fail.

mattaereal commented:
Feel free to apply any suggestion you think is appropriate that may be triggering style issues since it's totally irrelevant to how I express what is meaningful in this scenario.

Also, you can squash and merge by clicking a button on the GitHub UI.

matheusaaguiar commented on docs/control-structures.rst:
```diff
@@ -900,3 +913,4 @@ in scope in the block that follows.
     The caller always retains at least 1/64th of the gas in a call and thus
     even if the called contract goes out of gas, the caller still
     has some gas left.
+    
```
```suggestion
```
